File: src/rfq_test/crypto/eip712.py
"""EIP-712 v2 signing for RFQ quotes and conditional orders (TP/SL).

Mirrors the indexer's reference implementation at
`service/rfq/signature/eip712.go` byte-for-byte. The contract and indexer
verify v2 signatures by recomputing the same digest; any drift here will
cause `ErrSignatureMismatch`.

The protocol does NOT use `eth_signTypedData_v4`. It defines a custom
EIP-712 layout: a fixed `EIP712Domain(string,string,uint256,address)`
domain separator combined with a hand-rolled `SignQuote` /
`SignedTakerIntent` type hash. Encoding rules:

  * `string` and decimal fields -> `keccak256(utf8(s))` (32 bytes).
  * `address` -> 20 bytes from `bech32_to_evm`, left-padded to 32 bytes.
  * `uint8` / `uint32` / `uint64` -> big-endian, right-aligned in 32 bytes.
  * `SignQuote.bindingKind` is `1` when a taker address is present and `0`
    for blind quotes.
  * Fixed terminators on `SignedTakerIntent`:
      - `unfilledActionKind = 0` (none) and `unfilledActionPrice = "0"`.
      - `cid` is `keccak256("")` when null.
      - `allowedRelayer` is the zero address (32 zero bytes) when null.

Final digest = `keccak256(0x19 || 0x01 || domainSeparator || msgHash)`.
"""

from __future__ import annotations

from decimal import Decimal
from typing import Optional, Union

import bech32
from eth_account import Account
from eth_hash.auto import keccak

# --- Domain ------------------------------------------------------------

EIP712_DOMAIN_NAME = "RFQ"
EIP712_DOMAIN_VERSION = "1"
EIP712_DOMAIN_TYPE = (
    b"EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
)

# --- Type hashes (must match indexer) ----------------------------------

SIGN_QUOTE_TYPE = (
    b"SignQuote("
    b"uint64 evmChainId,string marketId,uint64 rfqId,address taker,uint8 takerDirection,"
    b"string takerMargin,string takerQuantity,address maker,"
    b"uint32 makerSubaccountNonce,string makerQuantity,string makerMargin,"
    b"string price,uint8 expiryKind,uint64 expiryValue,string minFillQuantity,"
    b"uint8 bindingKind"
    b")"
)

SIGNED_TAKER_INTENT_TYPE = (
    b"SignedTakerIntent("
    b"uint8 version,uint64 evmChainId,address taker,uint64 epoch,uint64 rfqId,string marketId,"
    b"uint32 subaccountNonce,uint64 laneVersion,uint64 deadlineMs,"
    b"uint8 direction,string quantity,string margin,string worstPrice,"
    b"string minTotalFillQuantity,uint8 triggerKind,string triggerPrice,"
    b"uint8 unfilledActionKind,string unfilledActionPrice,string cid,"
    b"address allowedRelayer"
    b")"
)

# --- Constants ---------------------------------------------------------

DIRECTION_LONG = 0
DIRECTION_SHORT = 1

EXPIRY_KIND_TIMESTAMP = 0
EXPIRY_KIND_HEIGHT = 1

TRIGGER_KIND_IMMEDIATE = 0
TRIGGER_KIND_MARK_PRICE_GTE = 1
TRIGGER_KIND_MARK_PRICE_LTE = 2

UNFILLED_ACTION_KIND_NONE = 0  # only kind that's part of the v2 digest

BINDING_KIND_BLIND = 0
BINDING_KIND_TAKER_SPECIFIC = 1


# --- Address helpers ---------------------------------------------------

def bech32_to_evm(addr: str) -> bytes:
    """Decode an `inj1...` bech32 address into its 20-byte EVM form.

    Mirrors `signature.AddrToEVM` in the indexer.
    """
    hrp, data = bech32.bech32_decode(addr)
    if hrp != "inj":
        raise ValueError(f"unexpected bech32 hrp {hrp!r}, want 'inj'")
    if data is None:
        raise ValueError(f"invalid bech32: {addr!r}")
    raw = bech32.convertbits(data, 5, 8, False)
    if raw is None or len(raw) != 20:
        raise ValueError(f"expected 20-byte address, got {len(raw) if raw else 0}")
    return bytes(raw)


# --- Word encoders (return 32 bytes each) ------------------------------

def _enc_u8(v: int) -> bytes:
    if not 0 <= v <= 0xFF:
        raise ValueError(f"uint8 out of range: {v}")
    return b"\x00" * 31 + bytes([v])


def _enc_u32(v: int) -> bytes:
    if not 0 <= v <= 0xFFFFFFFF:
        raise ValueError(f"uint32 out of range: {v}")
    return b"\x00" * 28 + v.to_bytes(4, "big")


def _enc_u64(v: int) -> bytes:
    if not 0 <= v <= 0xFFFFFFFFFFFFFFFF:
        raise ValueError(f"uint64 out of range: {v}")
    return b"\x00" * 24 + v.to_bytes(8, "big")


def _enc_addr(addr20: bytes) -> bytes:
    if len(addr20) != 20:
        raise ValueError(f"expected 20-byte address, got {len(addr20)}")
    return b"\x00" * 12 + addr20


def _enc_zero_addr() -> bytes:
    return b"\x00" * 32


def _enc_string(s: str) -> bytes:
    return keccak(s.encode("utf-8"))


def _enc_decimal(s: str) -> bytes:
    """Decimals are encoded as `keccak256(utf8(s))`. The string must be the
    same one sent on the wire (no formatting drift)."""
    return _enc_string(s)


def _enc_opt_string(s: Optional[str]) -> bytes:
    return _enc_string(s if s is not None else "")


def _enc_opt_addr(bech: Optional[str]) -> bytes:
    if not bech:
        return _enc_zero_addr()
    return _enc_addr(bech32_to_evm(bech))


def _direction_word(direction: str) -> int:
    d = direction.lower()
    if d == "long":
        return DIRECTION_LONG
    if d == "short":
        return DIRECTION_SHORT
    raise ValueError(f"invalid direction {direction!r}")


def _trigger_words(trigger_type: str, trigger_price: Optional[str]) -> tuple[int, str]:
    t = trigger_type.lower() if trigger_type else "immediate"
    if t == "immediate":
        return TRIGGER_KIND_IMMEDIATE, "0"
    if t == "mark_price_gte":
        if not trigger_price:
            raise ValueError("mark_price_gte trigger requires a price")
        return TRIGGER_KIND_MARK_PRICE_GTE, trigger_price
    if t == "mark_price_lte":
        if not trigger_price:
            raise ValueError("mark_price_lte trigger requires a price")
        return TRIGGER_KIND_MARK_PRICE_LTE, trigger_price
    raise ValueError(f"unknown trigger type {trigger_type!r}")


# --- Domain separator + final digest -----------------------------------

def domain_separator(evm_chain_id: int, verifying_contract_bech32: str) -> bytes:
    """Compute the EIP-712 domain separator for the RFQ contract."""
    addr20 = bech32_to_evm(verifying_contract_bech32)
    type_hash = keccak(EIP712_DOMAIN_TYPE)
    buf = b"".join((
        type_hash,
        _enc_string(EIP712_DOMAIN_NAME),
        _enc_string(EIP712_DOMAIN_VERSION),
        _enc_u64(evm_chain_id),  # chainId is uint256 in the type, but EVM chain
                                  # IDs fit in uint64 (Injective is 1439/1776/etc).
                                  # Indexer code uses big.Int.FillBytes(32) which
                                  # is byte-identical to right-aligned u64 here.
        _enc_addr(addr20),
    ))
    return keccak(buf)


def _final_digest(domain_sep: bytes, msg_hash: bytes) -> bytes:
    return keccak(b"\x19\x01" + domain_sep + msg_hash)


# --- Quote digest (SignQuote) ------------------------------------------

def sign_quote_digest(
    *,
    evm_chain_id: int,
    verifying_contract_bech32: str,
    market_id: str,
    rfq_id: int,
    taker_bech32: Optional[str],
    direction: str,
    taker_margin: str,
    taker_quantity: str,
    maker_bech32: str,
    maker_subaccount_nonce: int,
    maker_quantity: str,
    maker_margin: str,
    price: str,
    expiry_kind: int,           # 0 = timestamp_ms, 1 = block height
    expiry_value: int,
    min_fill_quantity: Optional[str] = None,
) -> bytes:
    """Compute the 32-byte EIP-712 v2 digest for a SignQuote."""
    type_hash = keccak(SIGN_QUOTE_TYPE)
    msg = b"".join((
        type_hash,
        _enc_u64(evm_chain_id),
        _enc_string(market_id),
        _enc_u64(rfq_id),
        _enc_opt_addr(taker_bech32),
        _enc_u8(_direction_word(direction)),
        _enc_decimal(taker_margin),
        _enc_decimal(taker_quantity),
        _enc_addr(bech32_to_evm(maker_bech32)),
        _enc_u32(maker_subaccount_nonce),
        _enc_decimal(maker_quantity),
        _enc_decimal(maker_margin),
        _enc_decimal(price),
        _enc_u8(expiry_kind),
        _enc_u64(expiry_value),
        _enc_decimal(min_fill_quantity if min_fill_quantity is not None else "0"),
        _enc_u8(
            BINDING_KIND_TAKER_SPECIFIC if taker_bech32 else BINDING_KIND_BLIND
        ),
    ))
    domain_sep = domain_separator(evm_chain_id, verifying_contract_bech32)
    return _final_digest(domain_sep, keccak(msg))


# --- Conditional order digest (SignedTakerIntent) ----------------------

def signed_taker_intent_digest(
    *,
    evm_chain_id: int,
    verifying_contract_bech32: str,
    version: int,
    taker_bech32: str,
    epoch: int,
    rfq_id: int,
    market_id: str,
    subaccount_nonce: int,
    lane_version: int,
    deadline_ms: int,
    direction: str,
    quantity: str,
    margin: str,
    worst_price: str,
    min_total_fill_quantity: str,
    trigger_type: str,
    trigger_price: Optional[str],
    cid: Optional[str] = None,
    allowed_relayer: Optional[str] = None,
) -> bytes:
    """Compute the 32-byte EIP-712 v2 digest for a SignedTakerIntent.

    `unfilled_action` is intentionally not a parameter — v2 digests fix
    `unfilledActionKind = 0` and `unfilledActionPrice = "0"`. The wire
    payload may still carry `unfilled_action` (the indexer uses it
    post-trigger), but it is not bound by the signature.
    """
    trigger_kind, encoded_trigger_price = _trigger_words(trigger_type, trigger_price)

    type_hash = keccak(SIGNED_TAKER_INTENT_TYPE)
    msg = b"".join((
        type_hash,
        _enc_u8(version),
        _enc_u64(evm_chain_id),
        _enc_addr(bech32_to_evm(taker_bech32)),
        _enc_u64(epoch),
        _enc_u64(rfq_id),
        _enc_string(market_id),
        _enc_u32(subaccount_nonce),
        _enc_u64(lane_version),
        _enc_u64(deadline_ms),
        _enc_u8(_direction_word(direction)),
        _enc_decimal(quantity),
        _enc_decimal(margin),
        _enc_decimal(worst_price),
        _enc_decimal(min_total_fill_quantity),
        _enc_u8(trigger_kind),
        _enc_decimal(encoded_trigger_price),
        _enc_u8(UNFILLED_ACTION_KIND_NONE),
        _enc_decimal("0"),
        _enc_opt_string(cid),
        _enc_opt_addr(allowed_relayer),
    ))
    domain_sep = domain_separator(evm_chain_id, verifying_contract_bech32)
    return _final_digest(domain_sep, keccak(msg))


# --- Signing -----------------------------------------------------------

def _normalize_pk(private_key: str) -> bytes:
    pk = private_key[2:] if private_key.startswith("0x") else private_key
    return bytes.fromhex(pk)


def _sign_digest(digest: bytes, private_key: str) -> str:
    """Sign a 32-byte digest with secp256k1 and return `0x` + r||s||v hex.

    `v` is serialized as compact y-parity (`0` or `1`), matching
    ws-client/feat/changes-for-eip-712.
    """
    if len(digest) != 32:
        raise ValueError(f"digest must be 32 bytes, got {len(digest)}")
    account = Account.from_key(_normalize_pk(private_key))
    sig = account.unsafe_sign_hash(digest)
    v = sig.v - 27 if sig.v >= 27 else sig.v
    sig_bytes = sig.r.to_bytes(32, "big") + sig.s.to_bytes(32, "big") + bytes([v])
    return "0x" + sig_bytes.hex()


def sign_quote_v2(
    *,
    private_key: str,
    evm_chain_id: int,
    verifying_contract_bech32: str,
    market_id: str,
    rfq_id: int,
    taker: Optional[str],
    direction: str,
    taker_margin: Union[str, Decimal],
    taker_quantity: Union[str, Decimal],
    maker: str,
    maker_margin: Union[str, Decimal],
    maker_quantity: Union[str, Decimal],
    price: Union[str, Decimal],
    expiry_ms: Optional[int] = None,
    expiry_height: Optional[int] = None,
    maker_subaccount_nonce: int = 0,
    min_fill_quantity: Optional[Union[str, Decimal]] = None,
) -> str:
    """Sign a quote with EIP-712 v2 and return `0x` + r||s||v hex.

    Pass exactly one of `expiry_ms` or `expiry_height`. All decimal fields
    must be the exact string sent on the wire (pre-quantized to the
    market tick size — see `utils.price.quantize_to_tick`).
    """
    if (expiry_ms is None) == (expiry_height is None):
        raise ValueError("provide exactly one of expiry_ms or expiry_height")

    if expiry_ms is not None:
        expiry_kind, expiry_value = EXPIRY_KIND_TIMESTAMP, int(expiry_ms)
    else:
        expiry_kind, expiry_value = EXPIRY_KIND_HEIGHT, int(expiry_height)

    digest = sign_quote_digest(
        evm_chain_id=evm_chain_id,
        verifying_contract_bech32=verifying_contract_bech32,
        market_id=market_id,
        rfq_id=int(rfq_id),
        taker_bech32=taker,
        direction=direction,
        taker_margin=str(taker_margin),
        taker_quantity=str(taker_quantity),
        maker_bech32=maker,
        maker_subaccount_nonce=int(maker_subaccount_nonce),
        maker_quantity=str(maker_quantity),
        maker_margin=str(maker_margin),
        price=str(price),
        expiry_kind=expiry_kind,
        expiry_value=expiry_value,
        min_fill_quantity=None if min_fill_quantity is None else str(min_fill_quantity),
    )
    return _sign_digest(digest, private_key)


def sign_conditional_order_v2(
    *,
    private_key: str,
    evm_chain_id: int,
    verifying_contract_bech32: str,
    version: int,
    taker: str,
    epoch: int,
    rfq_id: int,
    market_id: str,
    subaccount_nonce: int,
    lane_version: int,
    deadline_ms: int,
    direction: str,
    quantity: Union[str, Decimal],
    margin: Union[str, Decimal],
    worst_price: Union[str, Decimal],
    min_total_fill_quantity: Union[str, Decimal],
    trigger_type: str,
    trigger_price: Optional[Union[str, Decimal]] = None,
    cid: Optional[str] = None,
    allowed_relayer: Optional[str] = None,
) -> str:
    """Sign a conditional order (TP/SL) with EIP-712 v2.

    Returns `0x` + r||s||v hex.
    """
    digest = signed_taker_intent_digest(
        evm_chain_id=evm_chain_id,
        verifying_contract_bech32=verifying_contract_bech32,
        version=int(version),
        taker_bech32=taker,
        epoch=int(epoch),
        rfq_id=int(rfq_id),
        market_id=market_id,
        subaccount_nonce=int(subaccount_nonce),
        lane_version=int(lane_version),
        deadline_ms=int(deadline_ms),
        direction=direction,
        quantity=str(quantity),
        margin=str(margin),
        worst_price=str(worst_price),
        min_total_fill_quantity=str(min_total_fill_quantity),
        trigger_type=trigger_type,
        trigger_price=None if trigger_price is None else str(trigger_price),
        cid=cid,
        allowed_relayer=allowed_relayer,
    )
    return _sign_digest(digest, private_key)

File: examples/ts-mm/eip712.ts
/**
 * RFQ EIP-712 v2 signing — TypeScript reference.
 *
 * Mirrors the indexer's `SignQuote` typed-data layout byte-for-byte. The
 * protocol does NOT use `eth_signTypedData_v4`; it has a fixed
 * `EIP712Domain` and a hand-rolled `SignQuote` typeHash with each field
 * encoded into a 32-byte word. Decimal fields are hashed as
 * `keccak256(utf8(s))`, so the wire price MUST equal the signed price
 * byte-for-byte.
 *
 * Standalone — no rfq_test dependency. Partners can vendor this file.
 */

import { getBytes, hexlify, keccak256, SigningKey, toUtf8Bytes } from "ethers";

export interface SignQuoteV2Input {
  privateKey: string;          // 0x-prefixed or bare hex
  evmChainId: number;          // 1439 testnet, 1776 mainnet
  contractAddress: string;     // RFQ contract bech32, "inj1..."
  marketId: string;
  rfqId: number | bigint;
  taker?: string | null;       // bech32 inj1...; omit/null for blind quotes
  direction: "long" | "short";
  takerMargin: string;         // FPDecimal string, byte-identical to wire
  takerQuantity: string;
  maker: string;               // bech32 inj1...
  makerSubaccountNonce: number;
  makerQuantity: string;
  makerMargin: string;
  price: string;               // tick-quantized BEFORE signing
  expiryMs?: number | bigint;  // exactly one of expiryMs / expiryHeight
  expiryHeight?: number | bigint;
  minFillQuantity?: string;    // sent as "0" in the digest when omitted
}

export interface SignMakerChallengeV2Input {
  privateKey: string;          // 0x-prefixed or bare hex
  evmChainId: number;
  contractAddress: string;     // bech32
  maker: string;               // bech32 inj1...
  nonceHex: string;            // 32-byte hex, with or without 0x
  expiresAt: number | bigint;
}

const DOMAIN_TYPE =
  "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)";

const SIGN_QUOTE_TYPE =
  "SignQuote(string marketId,uint64 rfqId,address taker,uint8 takerDirection," +
  "string takerMargin,string takerQuantity,address maker,uint32 makerSubaccountNonce," +
  "string makerQuantity,string makerMargin,string price,uint8 expiryKind," +
  "uint64 expiryValue,string minFillQuantity,uint8 bindingKind)";

const STREAM_AUTH_CHALLENGE_TYPE =
  "StreamAuthChallenge(uint64 evmChainId,address maker,bytes32 nonce,uint64 expiresAt)";

// --- bech32 → 20-byte EVM address (HRP = "inj") ------------------------------
//
// Self-contained 5-bit→8-bit reconverter so we don't pull in a bech32 dep just
// for the inline reference. Consult BIP-0173 for the spec.

const BECH32_CHARSET = "qpzry9x8gf2tvdw0s3jn54khce6mua7l";

export function bech32ToEvm(addr: string): Uint8Array {
  const sep = addr.lastIndexOf("1");
  if (sep < 1 || sep > addr.length - 7) {
    throw new Error(`bad bech32 layout: ${addr}`);
  }
  const hrp = addr.slice(0, sep).toLowerCase();
  if (hrp !== "inj") {
    throw new Error(`expected inj address, got hrp=${hrp}`);
  }
  const dataPart = addr.slice(sep + 1).toLowerCase();
  const data: number[] = [];
  for (const c of dataPart) {
    const v = BECH32_CHARSET.indexOf(c);
    if (v === -1) throw new Error(`bad bech32 char: ${c}`);
    data.push(v);
  }
  // Strip 6-symbol checksum
  const fiveBit = data.slice(0, -6);

  // Re-pack 5-bit groups → 8-bit bytes (no padding)
  let acc = 0;
  let bits = 0;
  const out: number[] = [];
  for (const v of fiveBit) {
    acc = (acc << 5) | v;
    bits += 5;
    while (bits >= 8) {
      bits -= 8;
      out.push((acc >> bits) & 0xff);
    }
  }
  if (bits >= 5 || (acc & ((1 << bits) - 1)) !== 0) {
    throw new Error("bech32: trailing bits");
  }
  if (out.length !== 20) {
    throw new Error(`expected 20-byte address, got ${out.length}`);
  }
  return new Uint8Array(out);
}

// --- 32-byte word encoders --------------------------------------------------

function uintWord(value: bigint, byteWidth: number): Uint8Array {
  if (value < BigInt(0)) throw new Error(`negative uint: ${value}`);
  const out = new Uint8Array(32);
  let v = value;
  for (let i = 0; i < byteWidth; i++) {
    out[31 - i] = Number(v & BigInt(0xff));
    v >>= BigInt(8);
  }
  if (v !== BigInt(0)) throw new Error(`uint overflow for width ${byteWidth}: ${value}`);
  return out;
}

function stringWord(s: string): Uint8Array {
  return getBytes(keccak256(toUtf8Bytes(s)));
}

function addressWord(b20: Uint8Array): Uint8Array {
  if (b20.length !== 20) throw new Error("addressWord requires 20 bytes");
  const out = new Uint8Array(32);
  out.set(b20, 12);
  return out;
}

function zeroAddressWord(): Uint8Array {
  return new Uint8Array(32);
}

function concat(...parts: Uint8Array[]): Uint8Array {
  const total = parts.reduce((n, p) => n + p.length, 0);
  const out = new Uint8Array(total);
  let pos = 0;
  for (const p of parts) {
    out.set(p, pos);
    pos += p.length;
  }
  return out;
}

function domainSeparator(evmChainId: number, contractBech32: string): Uint8Array {
  return getBytes(
    keccak256(
      concat(
        stringWord(DOMAIN_TYPE),
        stringWord("RFQ"),
        stringWord("1"),
        uintWord(BigInt(evmChainId), 8),
        addressWord(bech32ToEvm(contractBech32)),
      ),
    ),
  );
}

/**
 * Sign an RFQ quote with EIP-712 v2.
 * Returns a 0x-prefixed 65-byte hex string (r ‖ s ‖ v).
 *
 * Pair this with `sign_mode: "v2"` on the wire payload — the indexer
 * rejects empty values with `value of message.sign_mode must be one of
 * "v1", "v2"`.
 */
export function signQuoteV2(input: SignQuoteV2Input): string {
  if ((input.expiryMs == null) === (input.expiryHeight == null)) {
    throw new Error("provide exactly one of expiryMs or expiryHeight");
  }
  const expiryKind: bigint = input.expiryMs != null ? BigInt(0) : BigInt(1);
  const expiryValue: bigint = BigInt((input.expiryMs ?? input.expiryHeight)!);
  const directionByte: bigint = input.direction === "long" ? BigInt(0) : BigInt(1);
  const mfq = input.minFillQuantity ?? "0";

  const msgHash = getBytes(
    keccak256(
      concat(
        stringWord(SIGN_QUOTE_TYPE),
        stringWord(input.marketId),
        uintWord(BigInt(input.rfqId), 8),
        input.taker ? addressWord(bech32ToEvm(input.taker)) : zeroAddressWord(),
        uintWord(directionByte, 1),
        stringWord(input.takerMargin),
        stringWord(input.takerQuantity),
        addressWord(bech32ToEvm(input.maker)),
        uintWord(BigInt(input.makerSubaccountNonce), 4),
        stringWord(input.makerQuantity),
        stringWord(input.makerMargin),
        stringWord(input.price),
        uintWord(expiryKind, 1),
        uintWord(expiryValue, 8),
        stringWord(mfq),
        uintWord(input.taker ? BigInt(1) : BigInt(0), 1),
      ),
    ),
  );

  const ds = domainSeparator(input.evmChainId, input.contractAddress);
  const digest = keccak256(concat(getBytes("0x1901"), ds, msgHash));

  const pkHex = "0x" + input.privateKey.replace(/^0x/, "");
  const key = new SigningKey(pkHex);
  const sig = key.sign(digest);
  const yParity = sig.yParity ?? (sig.v >= 27 ? sig.v - 27 : sig.v);
  return hexlify(
    concat(getBytes(sig.r), getBytes(sig.s), new Uint8Array([yParity & 1])),
  );
}

export function signMakerChallengeV2(input: SignMakerChallengeV2Input): string {
  const nonceHex = input.nonceHex.replace(/^0x/, "");
  if (nonceHex.length !== 64) {
    throw new Error(`expected 32-byte nonce hex, got ${nonceHex.length / 2} bytes`);
  }

  const msgHash = getBytes(
    keccak256(
      concat(
        stringWord(STREAM_AUTH_CHALLENGE_TYPE),
        uintWord(BigInt(input.evmChainId), 8),
        addressWord(bech32ToEvm(input.maker)),
        getBytes("0x" + nonceHex),
        uintWord(BigInt(input.expiresAt), 8),
      ),
    ),
  );

  const ds = domainSeparator(input.evmChainId, input.contractAddress);
  const digest = keccak256(concat(getBytes("0x1901"), ds, msgHash));

  const pkHex = "0x" + input.privateKey.replace(/^0x/, "");
  const key = new SigningKey(pkHex);
  const sig = key.sign(digest);
  const yParity = sig.yParity ?? (sig.v >= 27 ? sig.v - 27 : sig.v);
  return hexlify(
    concat(getBytes(sig.r), getBytes(sig.s), new Uint8Array([yParity & 1])),
  );
}
